fvm/src/syscalls/error.rs

```diff
@@ -50,59 +50,37 @@ impl Abort {
     }
 }
 
-/// Wraps an execution error in a Trap.
-impl From<Abort> for Trap {
-    fn from(a: Abort) -> Self {
-        Trap::from(Box::new(Envelope::wrap(a)) as Box<dyn std::error::Error + Send + Sync + 'static>)
-    }
-}
-
 /// Unwraps a trap error from an actor into an "abort".
-impl From<Trap> for Abort {
-    fn from(t: Trap) -> Self {
-        use std::error::Error;
+impl From<anyhow::Error> for Abort {
+    fn from(e: anyhow::Error) -> Self {
+        if let Some(trap) = e.downcast_ref::<Trap>() {
+            return match trap {
+                | Trap::MemoryOutOfBounds
+                | Trap::TableOutOfBounds
+                | Trap::IndirectCallToNull
+                | Trap::BadSignature
+                | Trap::IntegerOverflow
+                | Trap::IntegerDivisionByZero
+                | Trap::BadConversionToInteger
+                | Trap::UnreachableCodeReached
 
-        // Actor panic/wasm error.
-        if let Some(code) = t.trap_code() {
-            return Abort::Exit(
-                ExitCode::SYS_ILLEGAL_INSTRUCTION,
-                code.to_string(),
-                NO_DATA_BLOCK_ID,
-            );
-        }
-
-        // Try to get a smuggled error back.
-        t.source()
-            .and_then(|e| e.downcast_ref::<Envelope>())
-            .and_then(|e| e.take())
-            // Otherwise, treat this as a fatal error.
-            .unwrap_or_else(|| Abort::Fatal(t.into()))
-    }
-}
+                // Should require the atomic feature to be enabled, but we might as well just
+                // handle this.
+                | Trap::HeapMisaligned
+                | Trap::AtomicWaitNonSharedMemory
 
-/// A super special secret error type for stapling an error to a trap in a way that allows us to
-/// pull it back out.
-///
-/// BE VERY CAREFUL WITH THIS ERROR TYPE: Its source is self-referential.
-#[derive(Display, Debug)]
-#[display(fmt = "wrapping error")]
-struct Envelope {
-    inner: Mutex<Option<Abort>>,
-}
-
-impl Envelope {
-    fn wrap(a: Abort) -> Self {
-        Self {
-            inner: Mutex::new(Some(a)),
+                // I think this is fatal? But I'm not sure.
+                | Trap::StackOverflow => Abort::Exit(
+                    ExitCode::SYS_ILLEGAL_INSTRUCTION,
+                    trap.to_string(),
+                    NO_DATA_BLOCK_ID,
+                ),
+                _ => Abort::Fatal(anyhow!("unexpected wasmtime trap: {}", trap)),
```

Comment:
Nit: Do we actually want a default case here? I feel like it may be better to be very explicit here, and make things not build if we don't handle all traps explicitly.

Reply:
I'd love to... but the enum is marked as "non exhaustive".